generator/src/js/helpers/normalizever.js

```diff
@@ -0,0 +1,21 @@
+export default (version) => {
+  const letterRE = /([abcdefghijlmnopqrstuvwxyz]+$)/;
+
+  // convert to string, just in case
+  version = String(version);
+
+  // get the number of subversions, e.g. 1.0.1 = 2
+  const subversions = version.match(/\./g) === null ? 0 : version.match(/\./g).length;
+
+  // normalize based on the number of subversions
+  if (subversions === 0) {
+    version = `${version}.0.0`;  // 1 --> 1.0.0
+  } else if (subversions === 1) {
+    version = `${version.split('.')[0]}.0.${version.split('.')[1]}`;  // 1.1e --> 1.0.1e
```

Comment:
The library that it uses can only compare semantic versions:
https://semver.org/

Most of the server software generally follows this, but OpenSSL doesn't.

Reply:
Doesn't line 14 transform things that have only a major and minor version (e.g. `v1.3`) into semantic versions with a major minor and patch, moving the minor version to the patch and setting the new minor version to `0` (e.g. `v1.0.3`)?

Reply:
IMHO, you should not rely on semver library.
Instead, you should rely on a String that you will compare (now some projects use Codenames instead of semver).
Then the more powerfull approach may be to use a context and building the configuration regarding this context (e.g.: an apache server with a specific version and OpenSSL with another). I think it may be interesting to work on something like that, but there is a lot of work to do this.
There is an issue about this kind of need: https://github.com/mozilla/server-side-tls/issues/243

Reply:
I think @dangoncalves is calling out #243 because the Oracle HTTP Server also doesn't use semver (just as openssl doesn't use semver). For example the current version of OHS is [12.2.1.3](https://www.oracle.com/technetwork/middleware/webtier/downloads/index.html)

> you should rely on a String that you will compare 

@dangoncalves do you have a method in mind for how to do version comparisons for these non semver version schemes?

> you should not rely on semver library

wouldn't this still have value for the majority of products that this tool applies to which *do* use semver?

Reply:
Writing string comparisons is a) stupendously complicated and b) prone to breaking. I'd rather deal with the weird one-offs manually by making them into something that semver can understand that write an extremely brittle and complex version comparison library.

Reply:
For reference, the semver library is approximately 1500 lines of code. And while I could certainly do it in fewer lines of code (since there is stuff I don't need), I doubt I could do it nearly as well. And I would be stuck maintaining it in perpetuity.

Reply:
I'd recommend we
* use semver by default
* intentionally use application specific comparison method for non conforming applications (like openssl)

I realized that I think our general case which is supposed to deal with openssl [has a bug](https://github.com/mozilla/server-side-tls/pull/242#discussion_r280588245) and it may benefit us to make this not general (and openssl specific instead). This way later when we need to compare OHS for example we create an application specific method that normalizes OHS versions (possibly in the math way I [suggest below for openssl](https://github.com/mozilla/server-side-tls/pull/242#discussion_r280588245)

Reply:
What about regular expression? 
You can find documentation for javascript [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp)
And a tester [here](https://regex101.com/). If you want the source for this tester, there are [here](https://github.com/firasdib/Regex101/issues)

Reply:
My Bad... There is no source for the tester. You can use [this](https://github.com/gskinner/regexr/) instead

Reply:
Hey wait... @april you already opened an issue about that: #239

Reply:
Yes, every piece of software would have its own template, but not its own versioning code. That's how we ended up with the mess we have now.